### PR TITLE
Refactor quota grace: DRY violations, test gaps, and unmount race

### PR DESCRIPTION
**Pull request by chiedoclaude**
## Summary\n\n- Add `IsInQuotaGrace()` method on `Subscription` to replace the repeated 3-field check (`QuotaPlanID != nil && QuotaEntitlementsUntil != nil && time.Now().Before(...)`) used in 3 places\n- Extract `handleEndQuotaGraceNow` helper from `handleDowngrade` (~50 lines), also moves `buildDowngradeLimitWarnings` inside the transaction for consistency\n- Extract shared `LimitWarningsList` component from `DowngradeConfirmDialog` and `EndQuotaGraceConfirmDialog` to eliminate duplicated warning list rendering\n- Fix `EndQuotaGraceConfirmDialog` unmount race by always rendering it (controlled by `open` prop) instead of conditionally mounting based on quota data\n- Remove unused `effectiveLimits` prop from `PlanCard` (audit retention already correctly uses `plan.audit_retention_days`)\n- Add 409 conflict test for the expired quota grace TOCTOU edge case (double-call scenario)\n\nCloses #805\n\n## Test plan\n\n### Developer\n- [x] Backend tests pass (`go test ./api/... ./db/...`)\n- [x] Frontend tests pass (959 tests, `make test-frontend`)\n- [x] TypeScript compiles cleanly (`npx tsc --noEmit`)\n- [x] New test `TestDowngrade_EndQuotaGraceNow_DoubleCallReturns409` passes\n\n### Manual Testing\n- [ ] Verify audit retention days display correctly during quota grace period\n- [ ] Verify downgrade dialog doesn't unmount mid-display\n\nhttps://claude.ai/code/session_01DucJjcZmX1AfiJRxuCHV3T

### CODE REVIEW

File: api/billing.go [hunk 1]
```diff
@@ -761,6 +707,67 @@ func handleDowngrade(deps *Deps) http.HandlerFunc {
...
+		defer db.RollbackTx(r.Context(), tx)
+	}
+
+	warnings, warnErr := buildDowngradeLimitWarnings(r.Context(), deps, userID, freePlan)
+	if warnErr != nil {
+		log.Printf("[%s] Downgrade (end grace): limit warnings: %v", TraceID(r.Context()), warnErr)
+		CaptureError(r.Context(), warnErr)
+		RespondError(w, r, http.StatusInternalServerError, InternalError("Failed to verify plan limits"))
+		return
+	}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply api/billing.go]: Good catch — moved `buildDowngradeLimitWarnings` back before `BeginOrContinue` to match the `handleDowngrade` pattern. The function reads from `deps.DB` (not the tx), so holding the tx open during those queries was unnecessary. Fixed in 3cb91e1.

File: frontend/src/pages/billing/PlanDetailsCard.tsx [hunk 2]
```diff
@@ -193,17 +193,15 @@ function DowngradeSection({ usage, subscription }: DowngradeSectionProps) {
...
-      )}
+      <EndQuotaGraceConfirmDialog
+        open={showEndGraceConfirm}
+        onOpenChange={setShowEndGraceConfirm}
+        onConfirm={handleDowngrade}
+        isPending={isDowngrading}
+        error={downgradeError}
+        usage={usage}
+        paidEntitlementsEndDate={subscription.quota_entitlements_until ? formatDate(subscription.quota_entitlements_until) : ""}
+      />
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply frontend/src/pages/billing/PlanDetailsCard.tsx:204]: Agreed — using `"your grace period end"` as the fallback instead of an empty string. The dialog is normally gated by `showEndGraceConfirm` state, but the defensive fallback handles the edge case where data refetches mid-display. Fixed in 3cb91e1.

File: db/subscriptions.go [hunk 3]
```diff
@@ -441,7 +447,7 @@ func GetSubscriptionWithPlan(ctx context.Context, db DBTX, userID string) (*Subs
 	if sub == nil {
 		return nil, nil
 	}
-	if sub.QuotaPlanID != nil && sub.QuotaEntitlementsUntil != nil && !time.Now().Before(*sub.QuotaEntitlementsUntil) {
+	if sub.QuotaPlanID != nil && sub.QuotaEntitlementsUntil != nil && !sub.IsInQuotaGrace() {
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply db/subscriptions.go]: Good point — simplified to `!sub.IsInQuotaGrace() && sub.QuotaPlanID != nil`. The `ClearExpiredSubscriptionQuotaGrace` SQL has its own WHERE clause checking both fields, so the Go-side guard just needs to know there are quota columns worth attempting to clear. Fixed in 3cb91e1.

**Comment by chiedoclaude** [comment]
## 🤖 Watch Session Summary

### Changes Made
No changes made during this session.

### Decision Log

#### ✅ Implemented
- greptile-apps[bot] asked for Move buildDowngradeLimitWarnings before BeginOrContinue to avoid holding tx open unnecessarily → implemented in `3cb91e1`
- greptile-apps[bot] asked for Use non-empty fallback for paidEntitlementsEndDate when quota_entitlements_until is null → implemented in `3cb91e1`
- greptile-apps[bot] asked for Simplify redundant nil checks in GetSubscriptionWithPlan now that IsInQuotaGrace encapsulates them → implemented in `3cb91e1`

---
*Watch session ended after 6 minutes of inactivity. Processed 3 comments across 6 poll cycles.*

**Comment by chiedoclaude** [comment]
⚠️ **Auto-merge failed.** The `--automerge` flag was set, but the merge could not be completed. Please merge manually.
